program/src/processor.rs

```diff
@@ -644,3 +666,189 @@ pub fn process(program_id: &Pubkey, accounts: &[AccountInfo], input: &[u8]) -> P
+            // calculation should never return an error, hence the
+            // `unwrap` here.
+            //
+            // The scaling to `u128` prevents multiplication from breaking
+            // the `u64::MAX` ceiling, and the `token_supply == 0` check
+            // prevents `checked_div` returning `None` from a zero
+            // denominator.
+            let result = calculate_rewards_per_token(rewards, token_supply).unwrap();
+            // Evaluate.
+            if token_supply == 0 {
+                prop_assert_eq!(result, 0);
+            } else {
+                let expected = (rewards as u128)
+                    .checked_mul(REWARDS_PER_TOKEN_SCALING_FACTOR)
+                    .and_then(|product| product.checked_div(token_supply as u128))
+                    .unwrap();
+                prop_assert_eq!(result, expected);
+            }
+        }
+    }
+
+    // Since the marginal rate (current - last) within this function is describing
+    // rewards _per token_, which is calculated by `calculate_rewards_per_token`,
+    // the maximum permitted true reward difference on the system that can be
+    // tolerated by a maximum token account balance (100% of token supply) depends
+    // on the token supply of the mint.
+    //
+    // Consider the calculations for both functions:
+    //
+    //     rewards_per_token = (reward * 1e9) / mint_supply
+    //     eligible_rewards = (marginal_rewards_per_token * balance) / 1e9
+    //
+    // By rearranging the first formula (for `calculate_rewards_per_token`), we
+    // can calculate the total accumulated system-level reward based on the
+    // given rewards per token and mint supply.
+    //
+    //     accumulated_rewards =
+    //              (accumulated_rewards_per_token * mint_supply) / e9
+    //
+    // By rearranging the second formula (which is for this function
+    // `calculate_eligible_rewards`), and plugging in `u64::MAX * 1e9` for
+    // `eligible_rewards` (function upper bound), we can obtain a formula for
+    // the maximum marginal rate (current - last) supported.
+    //
+    //     max_marginal_rewards_per_token = (u64::MAX * 1e9) / balance
+    //
+    // The largest token account balance possible is 100% of the mint supply,
+    // so we can rewrite the rearranged second formula as follows.
+    //
+    //     max_marginal_rewards_per_token = (u64::MAX * 1e9) / mint_supply
+    //
+    // Now, plugging the formula for `max_marginal_rewards_per_token` into the
+    // `accumulated_rewards_per_token` of the rearranged first formula, we get
+    // the following.
+    //
+    //     max_accumulated_rewards = (
+    //                          ((u64::MAX * 1e9) / mint_supply) * mint_supply
+    //                        ) / 1e9
+    //
+    // As you can see, the mint supply can be factored out, and the max
+    // accumulated system-level reward can now be computed directly.
+    //
+    //     max_accumulated_rewards = u64::MAX
```

Comment:
Yes but the key here is that, without our formulae here, `max_accumulated_rewards` can be over `u64::MAX` since you could just deposit rewards of `u64::MAX` over and over again, applying the calculation for `rewards_per_token` and adding to the `u128`.

Reply:
Maybe I should make it clear that this value isn't tracked as-is in the program, but only in this arithmetic?